be/src/olap/rowset/segment_v2/binary_dict_page.cpp

```diff
@@ -0,0 +1,231 @@
+    size_t add_count = 1;
+    // here do not check is_page_full of dict_builder
+    // because it is checked in add
+    for (auto& dict_item: _dict_items) {
+        RETURN_IF_ERROR(_dict_builder->add(reinterpret_cast<const uint8_t*>(&dict_item), &add_count));
+    }
+    *dictionary_page = _dict_builder->finish();
+    _dict_builder->release();
+    _dict_items.clear();
+    return Status::OK();
+}
+
+BinaryDictPageDecoder::BinaryDictPageDecoder(Slice data, const PageDecoderOptions& options) :
+    _data(data),
+    _options(options),
+    _data_page_decoder(nullptr),
+    _dict_decoder(options.dict_decoder),
+    _parsed(false),
+    _encoding_type(UNKNOWN_ENCODING) { }
+
+Status BinaryDictPageDecoder::init() {
+    CHECK(!_parsed);
+    if (_data.size < BINARY_DICT_PAGE_HEADER_SIZE) {
+        LOG(WARNING) << "corrupted data, data size:" << _data.size;
+        return Status::Corruption(Substitute("invalid data size:$0, header size:$1",
+                _data.size, BINARY_DICT_PAGE_HEADER_SIZE));
+    }
+    size_t type = decode_fixed32_le((const uint8_t*)&_data.data[0]);
+    _encoding_type = static_cast<EncodingTypePB>(type);
+    _data.remove_prefix(BINARY_DICT_PAGE_HEADER_SIZE);
+    if (_encoding_type == DICT_ENCODING) {
+        DCHECK(_dict_decoder != nullptr) << "dict decoder pointer is nullptr";
+        _data_page_decoder.reset(new BitShufflePageDecoder<OLAP_FIELD_TYPE_INT>(_data, _options));
+    } else if (_encoding_type == PLAIN_ENCODING) {
+        DCHECK_EQ(_encoding_type, PLAIN_ENCODING);
+        // use plain page decoder to decode data
+        _data_page_decoder.reset(new BinaryPlainPageDecoder(_data, _options));
+    } else {
+        LOG(WARNING) << "invalide encoding type:" << _encoding_type;
+        return Status::Corruption(Substitute("invalid encoding type:$0", _encoding_type));
+    }
+
+    Status status = _data_page_decoder->init();
+    if (!status.ok()) {
+        LOG(WARNING) << "status:" << status.to_string();
+        return status;
+    }
+    _parsed = true;
+    return Status::OK();
+}
+
+Status BinaryDictPageDecoder::seek_to_position_in_page(size_t pos) {
+    return _data_page_decoder->seek_to_position_in_page(pos);
+}
+
+Status BinaryDictPageDecoder::next_batch(size_t* n, ColumnBlockView* dst) {
+    if (_encoding_type == PLAIN_ENCODING) {
+        return _data_page_decoder->next_batch(n, dst);
+    } else {
+        // dictionary encoding
+        DCHECK(_parsed);
+        if (PREDICT_FALSE(*n == 0)) {
+            *n = 0;
+            return Status::OK();
+        }
+        Slice* out = reinterpret_cast<Slice*>(dst->data());
+        _code_buf.resize((*n) * sizeof(int32_t));
+        
+        // copy the codewords into a temporary buffer first
+        // And then copy the strings corresponding to the codewords to the destination buffer
+        BitShufflePageDecoder<OLAP_FIELD_TYPE_INT>* data_ptr =
+                down_cast<BitShufflePageDecoder<OLAP_FIELD_TYPE_INT>*>(_data_page_decoder.get());
```

Comment:
Why do we need the raw pointer here? Couldn't we simply use _data_page_decoder->next_batch in the following code?

Reply:
my mistake. I will change it.